### DISCUSSION TITLE
Question about focus highlighting and FocusAdorner

### DISCUSSION THREAD

**Discussion by iksi4prs** [AvaloniaUI/Avalonia#10644] [Q&A] [answered]
Hi,
I've created a simple dialog, with 2 buttons, and selected the focused button using 
`button.AttachedToVisualTree += (s, e) => button.Focus();`
I was able to confirm that button is in focus, by changing the border to green using selector (see attached image).
But when I press the TAB key, then it looks like the "i'm in focus" visual indication, is changed.
(namely overridden by the framework, not done from my code)
I assume (but maybe wrong), its done by "FocusAdorner" (?)
Is there any way to control this ?
And why in the first place, the focus button is shown differently, untill press the TAB ?
Is the way to move to this state, in other way than "Focus()" ?
Thanks for your help,
Si.

![image](https://user-images.githubusercontent.com/65832579/224672143-19bc8de2-25d8-4c63-89a1-a0e439f4a26f.png)

**Comment by timunie** [comment, marked as answer]
Depends on the Avalonia version you use. In 11.0 you can define a Style for it like shown here: https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Themes.Fluent/Controls/AdornerLayer.xaml

**Reply by iksi4prs** [reply]
@timunie , 10x for your help ! (again).

Comparing to other UI systems, in conjunction to TAB navigation:
Why the "focus adorner" is not doing the paint, until I navigate using TAB ?
As I wrote, using the selector, I see the button is focused. 
So it looks like there are 2 focus states, with "regular focus" and one "super focus via TAB", which also uses the adorner.

In best scenario, I don't want to paint the focus myself at all.
Using the green border was just a result trying to figure out is the button at focus all,
As I didn't manage to cause the adorner to paint, before I navigated using the TAB.

**Reply by timunie** [reply]
I think it's the Keyboard-focus which is actually similar to what WPF does. Can you try to do this?

```cs
Dispatcher.UIThread.Post(() => KeyboardDevice.Instance.SetFocusedElement(myButton, NavigationMethod.Tab, KeyModifiers.None), 
  DispatcherPriority.Loaded);
```

**Reply by iksi4prs** [reply]
@timunie , worked like a charm.
I still don't understand one thing:
After I call button.Focus(),  pressing SPACE,  invokes the "click" of this button.
Doesn't it means the button also has the keyboard focus in addition to the logical focus ?

**Reply by timunie** [reply]
Hmm I'm not able to explain the exact implemention details. Sorry.

**Reply by iksi4prs** [reply]
@timunie , I will try to check the source code when I have some more time. 10x for your help !

**Reply by IanRawley** [reply]
> I think it's the Keyboard-focus which is actually similar to what WPF does. Can you try to do this?
> 
> ```cs
> Dispatcher.UIThread.Post(() => KeyboardDevice.Instance.SetFocusedElement(myButton, NavigationMethod.Tab, KeyModifiers.None), 
>   DispatcherPriority.Loaded);
> ```

How do we do this now that KeyboardDevice.Instance is internal?  <control>.Focus(NavigationMethod.Tab) doesn't seem to be working for me.

**Reply by samtun** [reply]
@IanRawley did you find a solution? I'm wondering the same thing.
